docs/user_guide/ml.md

```diff
@@ -1,8 +1,264 @@
 # Machine Learning
 
-In this section, the users will learn how to implement machine learning in their analysis with
-columnflow.
+In this section, the users will learn how to implement machine learning in their analysis with columnflow.
 
-Note: which tasks, how to trigger them, how to train, test and plot networks and outputs
+# How training happens in Columnflow: K-fold cross validation
+Machine learning in columnflow is implemented in a way that k-fold cross validation is enabled by default.
+In k-fold cross validation the dataset is split in k-parts of equal size.
+For each training, k-1 parts are used for the actual training and the remaining part is used for validation of the model.
+This process is repeated k-times, resulting in the training of k-model instances.
+In the end of the training columnflow will save all k-models, which are then usable for evaluation.
 
-TODO
+# Configure your custom machine learning class:
+To create a custom machine learning (ML) class in columnflow, it is imperative to inherit from the {py:class}`~columnflow.ml.MLModel` class.
+This inheritance ensures the availability of functions to manage and access config and model instances, as well as the necessary producers.
+The name of your custom ML class can be arbitrary, since `law` accesses your machine learning model using a `cls_name` in {py::meth}`~columnflow.util.DerivableMeta.derive`, e.g.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: test_model
+:end-at: test_model
+```
+The second argument in {py::meth}`~columnflow.util.DerivableMeta.derive` is a `cls_dict` which configures your subclass.
+The `cls_dict` needs to be flat.
+The keys of the dictionary are set as class attributes and are therefore also accessible by using `self`.
+The configuration with `derive` has two main advantages:
+- ==manageability==, since the dictionary can come from loading a config file and these can be changed fairly easy
+- ==flexibility==, multiple settings require only different configuration files
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: hyperparameters =
+:end-at: test_model = TestModel.derive(
+```
+One can also simply define class variables within the model.
+This is can be useful for attributes that don't change often, for example to define the datasets your model uses.
+Since these are also class attributes, they are accessible by using `self` and also shared between all instances of this model.
```

Comment:
```suggestion
Since these are also class attributes, they are accessible by using `self` and are also shared between all instances of this model.
```